pkg/api/publishWorker.go

```diff
@@ -0,0 +1,147 @@
+type PublishWorker struct {
+	ctx          context.Context
+	log          *zap.Logger
+	listener     <-chan []queries.StagedOriginatorEnvelope
+	notifier     chan<- bool
+	registrant   *registrant.Registrant
+	store        *sql.DB
+	subscription db.DBSubscription[queries.StagedOriginatorEnvelope]
+}
+
+func StartPublishWorker(
+	ctx context.Context,
+	log *zap.Logger,
+	reg *registrant.Registrant,
+	store *sql.DB,
+) (*PublishWorker, error) {
+	query := func(ctx context.Context, lastSeenID int64, numRows int32) ([]queries.StagedOriginatorEnvelope, int64, error) {
+		results, err := queries.New(store).SelectStagedOriginatorEnvelopes(
+			ctx,
+			queries.SelectStagedOriginatorEnvelopesParams{
+				LastSeenID: lastSeenID,
+				NumRows:    numRows,
+			},
+		)
+		if err != nil {
+			return nil, 0, err
+		}
+		if len(results) > 0 {
+			lastSeenID = results[len(results)-1].ID
+		}
+		return results, lastSeenID, nil
+	}
+	notifier := make(chan bool, 1)
+	subscription := db.NewDBSubscription(
+		ctx,
+		log,
+		query,
+		0, // lastSeenID
+		db.PollingOptions{Interval: time.Second, Notifier: notifier, NumRows: 100},
+	)
+	listener, err := subscription.Start()
+	if err != nil {
+		return nil, err
+	}
+
+	worker := &PublishWorker{
+		ctx:          ctx,
+		log:          log,
+		notifier:     notifier,
+		subscription: *subscription,
+		listener:     listener,
+		registrant:   reg,
+		store:        store,
+	}
+	go worker.start()
+
+	return worker, nil
+}
+
+func (p *PublishWorker) NotifyStagedPublish() {
+	select {
+	case p.notifier <- true:
+	default:
+	}
+}
+
+func (p *PublishWorker) start() {
+	for {
+		select {
+		case <-p.ctx.Done():
+			close(p.notifier)
+			return
+		case new_batch := <-p.listener:
+			for _, stagedEnv := range new_batch {
+				for !p.publishStagedEnvelope(stagedEnv) {
+					// Infinite retry on failure to publish; we cannot
+					// continue to the next envelope until this one is processed
+					time.Sleep(time.Second)
+				}
+			}
+		}
+	}
+}
+
+func (p *PublishWorker) publishStagedEnvelope(stagedEnv queries.StagedOriginatorEnvelope) bool {
+	logger := p.log.With(zap.Int64("sequenceID", stagedEnv.ID))
+	originatorEnv, err := p.registrant.SignStagedEnvelope(stagedEnv)
+	if err != nil {
+		logger.Error(
+			"Failed to sign staged envelope",
+			zap.Error(err),
+		)
+		return false
+	}
+	originatorBytes, err := proto.Marshal(originatorEnv)
+	if err != nil {
+		logger.Error("Failed to marshal originator envelope", zap.Error(err))
+		return false
+	}
+
+	q := queries.New(p.store)
+
+	// On unique constraint conflicts, no error is thrown, but numRows is 0
+	inserted, err := q.InsertGatewayEnvelope(
+		p.ctx,
+		queries.InsertGatewayEnvelopeParams{
+			OriginatorID:         int32(p.registrant.NodeID()),
+			OriginatorSequenceID: stagedEnv.ID,
+			Topic:                stagedEnv.Topic,
+			OriginatorEnvelope:   originatorBytes,
+		},
+	)
+	if err != nil {
+		logger.Error("Failed to insert gateway envelope", zap.Error(err))
+		return false
```

Comment:
There is a risk here that we infinitely retry on this step or one of the steps above in an unanticipated error case. I think that we should be okay here but a second pair of eyes can't hurt

Reply:
I don't know why the proto marshaling would fail, but if it did we would be in an endless loop of retries here.

Would probably have to be a maliciously crafted payload

Reply:
From my research, it seems like it fails if some expected fields of the proto are not set. In this case, the proto is `OriginatorEnvelope` with two fields that we are setting [here](https://github.com/xmtp/xmtpd/blob/a8cef43a30f683da94f79e36cb71f81225fe85bd/pkg/registrant/registrant.go#L93), so I expect that we should be able to prevent this error happening.

I still think we should catch the error just in case though, firstly so that we can detect it, secondly so that we don't screw up payload ordering if there is an unanticipated error type that is super common.